server/reload.go

```diff
@@ -846,6 +884,49 @@ func (s *Server) applyOptions(ctx *reloadContext, opts []option) {
 	s.Noticef("Reloaded server configuration")
 }
 
+// Update all cached debug and trace settings for every client
+func (s *Server) reloadClientTraceLevel() {
+	opts := s.getOpts()
+
+	if opts.NoLog {
+		return
+	}
+
+	update := func(c *client) {
+		// client.trace is commonly read while holding the lock
+		c.mu.Lock()
+		defer c.mu.Unlock()
+		c.setTraceLevel()
+	}
+
+	// Iterate over every client and update
+	// If this is not timely AND unexpected during a reload (with changes in trace level):
+	// Consider storing clients in a list, ten iterate without holding locks
+	// Or usage of sync.Map to store clients in the first place
+
+	if s.eventsEnabled() {
+		update(s.sys.client)
+	}
+
+	s.mu.Lock()
+	cMaps := []map[uint64]*client{s.clients, s.grTmpClients, s.routes, s.leafs}
+	for _, m := range cMaps {
+		for _, c := range m {
+			update(c)
+		}
+	}
+	s.mu.Unlock()
+
+	s.gateway.RLock()
```

Comment:
Feels like we need a write lock since we call something called update.

Reply:
@kozlovic correct me if I'm wrong, but this protects reading of the client. Writing the client is protected by its lock.

Reply:
Yes, correct. The RLock here is to protect going through the list of clients for the gateway, but then each client is updated using the client lock.